### PR TITLE
New ap iintegration

### PR DESCRIPTION
**Pull request by AGFarhat**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,6 +10,8 @@
 		"axios": "^0.16.2",
 		"react": "16.0.0-beta.5",
 		"react-native": "0.49.3",
+		"react-native-background-task": "^0.2.1",
```
Comment (rich521) [inline package.json]: remove all ^ from package json

File: src/App.js [hunk 2]
```diff
@@ -1,25 +1,155 @@
 import React, { Component } from 'react';
-import { Text, View } from 'react-native';
+import { Text, View, AppState } from 'react-native';
 import { Actions } from 'react-native-router-flux';
+import PushController from './components/PushController';
+import PushNotification from 'react-native-push-notification';
 import { Spinner, Button } from './components/common';
+// import BackgroundJob from 'react-native-background-job';
```
Comment (rich521) [inline src/App.js]: remove this

File: src/App.js [hunk 3]
```diff
@@ -1,25 +1,155 @@
...
+import BackgroundTask from 'react-native-background-task';
+
 // import { PermissionsAndroid } from 'react-native';
 import utils from './utils/methods';
-import defaultState from './utils/constants';
+
+// const firstJobKey = "firstJobKey";
+//onst regularJobKey = "regularJobKey";
```
Comment (rich521) [inline src/App.js]: remove this

File: src/App.js [hunk 4]
```diff
@@ -1,25 +1,155 @@
...
+//     //if (PushNotification) PushNotification.cancelAllLocalNotifications();
+//     BackgroundJob.cancelAll();
+//     BackgroundJob.schedule({
+//       jobKey: regularJobKey,
+//       period: 100000, //5mins
+//       //period: 86400000, //24hrs
+//       //networkType: 1,
+//     });
+//     }
+// });
```
Comment (rich521) [inline src/App.js]: remove all comments

File: src/App.js [hunk 5]
```diff
@@ -1,25 +1,155 @@
...
+  }
 
+  componentWillUnmount(){
+    AppState.removeEventListener('change', this.handleAppStateChange);
+  }
   componentWillMount() {
-    //utils.deleteLocalData();
-    this.fetchWeather();
+    //utils.deleteLocalData('@localStore');
+    //BackgroundJob.cancelAll();
```
Comment (rich521) [inline src/App.js]: here

File: src/Settings.js [hunk 6]
```diff
@@ -1,43 +1,87 @@
...
-          <CardSection>
-            <Text style = { styles.textStyle }>Settings Here</Text>
-          </CardSection>
-          <CardSection>
-            <Text style = { styles.textStyle }>Settings Here</Text>
-          </CardSection>
-        </Card>
+const initDate = new Date();
+initDate.setHours('07');
+initDate.setMinutes('00');
```
Comment (rich521) [inline src/Settings.js]: set constants for it at the top. 
```javascript
const initTime = { 
  hour: '07',
  mins: '00',
};
initDate.setHours(initTime.hours); 

```

File: src/App.js [hunk 7]
```diff
@@ -1,25 +1,155 @@
...
+
 // import { PermissionsAndroid } from 'react-native';
 import utils from './utils/methods';
-import defaultState from './utils/constants';
+
+// const firstJobKey = "firstJobKey";
+//onst regularJobKey = "regularJobKey";
+BackgroundTask.define(async () => {
+  const ID = 'ID: ' + Math.random().toFixed(4);
+  console.log(`1- ${ID} - ${new Date()} Background Job fired!.`);
```
Comment (rich521) [inline src/App.js]: remove comment

File: src/App.js [hunk 8]
```diff
@@ -1,25 +1,155 @@
...
+  PushNotification.localNotification({
+    title: "",
+    message: `weather: ${decision.weather.main.temp}
+    ${decision.weather.weather[0].main}
+    ${new Date()}`, // (required)
+    playSound: false,
+  });
+
+  BackgroundTask.schedule({
+      period: 900, //24 hrs
```
Comment (rich521) [inline src/App.js]: what is this number? put it at the top ;)

File: src/App.js [hunk 9]
```diff
@@ -1,25 +1,155 @@
...
   }
 
   fetchWeather = () => {
     this.setState({ remark: false, isFetching: true });
-    utils.getCachedItems().then(data => this.setState({ ...data, isFetching: false }));
+    utils.refreshCachedItems().then(data => {
+      this.setState({ ...data, isFetching: false });
+      //this.setState({ isRaining: true });
+      return this.state.isRaining;
+      //this.setPushNotification(new Date()); // TODO change this setDate
```
Comment (rich521) [inline src/App.js]: what is this comment for?

File: src/App.js [hunk 10]
```diff
@@ -1,25 +1,155 @@
...
+      //BackgroundJob.cancelAll();
+    }
   }
 
   fetchWeather = () => {
     this.setState({ remark: false, isFetching: true });
-    utils.getCachedItems().then(data => this.setState({ ...data, isFetching: false }));
+    utils.refreshCachedItems().then(data => {
+      this.setState({ ...data, isFetching: false });
+      //this.setState({ isRaining: true });
```
Comment (rich521) [inline src/App.js]: do we need setState ```isRaining```?

File: src/App.js [hunk 11]
```diff
@@ -1,25 +1,155 @@
...
+        period: (difference.getHours()*60*60 + difference.getMinutes()*60)*1000, //calculate time to set
+      });
+      */
+    }
+    if (appState === 'active') {
+      BackgroundTask.cancel();
+      // utils.getCachedItems().then(data => {
+      //   this.setState({ ...data });
+      //   console.log("recieved cached items");
+      // });
```
Comment (rich521) [inline src/App.js]: do we need this code?

File: src/App.js [hunk 12]
```diff
@@ -1,25 +1,155 @@
...
+      isFetching:false,
+      isRaining: false,
+      remark:'',
+      /* localStorage */
+      position: null,
+      weather: null,
+      lastUpdated: null,
+      /* SettingStorage */
+      reminderOn: false,
+      date: new Date('2017-01-01T07:00:00.000Z'),
```
Comment (rich521) [inline src/App.js]: do we need this string inside? new Date()?

File: src/App.js (hunk 11 above)
Comment (AGFarhat) [reply src/App.js]: we do, should be uncommented

File: src/App.js (hunk 10 above)
Comment (AGFarhat) [reply src/App.js]: yes

File: src/components/PushController.js [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import React, { Component } from 'react';
+import PushNotification from 'react-native-push-notification';
+
+export default class PushController extends Component{
+  componentDidMount(){
+    PushNotification.configure({
+      onNotification: function(notification) {
+        console.log( 'NOTIFICATION:', notification );
```
Comment (rich521) [inline src/components/PushController.js]: remove console

File: src/App.js [hunk 14]
```diff
@@ -1,25 +1,155 @@
...
   componentWillMount() {
-    //utils.deleteLocalData();
-    this.fetchWeather();
+    //utils.deleteLocalData('@localStore');
+    //BackgroundJob.cancelAll();
+    utils.getCachedItems().then(data => {
+      this.setState({ ...data , remark:true});
+    });
+
+    utils.setLocalData('@localStore', { description:'', isRaining:false});
```
Comment (rich521) [inline src/App.js]: use the KEY constant from method.js --> export KEY

File: src/App.js [hunk 15]
```diff
@@ -1,25 +1,155 @@
...
+    });
+  }
+  componentDidMount() {
+     AppState.addEventListener('change', this.handleAppStateChange);
+  }
+
+  handleAppStateChange = async (appState) => {
+    if (appState === 'background') {
+      this.setState({ isRaining: false });
+      const arabicFood = await utils.fetchSettings();
```
Comment (rich521) [inline src/App.js]: lets change this name

File: src/App.js [hunk 16]
```diff
@@ -1,25 +1,155 @@
...
+      const difference = new Date(new Date(arabicFood.date) - Date.now());
+      BackgroundTask.schedule({
+          period: (difference.getHours()*60*60 + difference.getMinutes()*60), //calculate time to set
```
Comment (rich521) [inline src/App.js]: can we store this a variable and use a meaningful name to describe the time?

File: src/App.js [hunk 17]
```diff
@@ -1,25 +1,155 @@
...
+      });
+      /*
+      BackgroundJob.schedule({
+        jobKey: firstJobKey,
+        period: (difference.getHours()*60*60 + difference.getMinutes()*60)*1000, //calculate time to set
+      });
+      */
```
Comment (rich521) [inline src/App.js]: remove this commnt

File: src/App.js (hunk 12 above)
Comment (AGFarhat) [reply src/App.js]: yep

File: src/utils/methods.js [hunk 18]
```diff
@@ -1,90 +1,125 @@
...
-    const URL_BASE = 'http://api.openweathermap.org/data/2.5/weather?';
+  getCurrentWeather: async ({latitude,longitude}) => {
+    // const URL_BASE = 'http://api.openweathermap.org/data/2.5/weather?';
+    const URL_BASE = 'http://api.openweathermap.org/data/2.5/forecast?';
     //const url = 'http://samples.openweathermap.org/data/2.5/weather?lat=35&lon=139&appid=b1b15e88fa797225412429c1c50c122a1';
     const url = `${URL_BASE}lat=${latitude}&lon=${longitude}&appid=${API}&units=metric`;
-
     return new Promise((resolve) => {
       axios.get(url).then(response => {
+        console.log(response.data);
```
Comment (rich521) [inline src/utils/methods.js]: remove this

File: src/utils/methods.js [hunk 19]
```diff
@@ -1,90 +1,125 @@
...
-      };
-      await utils.setLocalData(KEY.WEATHER, weatherData);
-      return weatherData;
-    }
-
-    let { position, weather, lastUpdated } = JSON.parse(localStore);
-
-    if (utils.getCurrentTime() - new Date(lastUpdated) > REFRESH_TIME){ //refresh time limit
+    let { position, weather, lastUpdated, isRaining, description } = await utils.getCachedItems(); //get local data
+    console.log(lastUpdated);
```
Comment (rich521) [inline src/utils/methods.js]: remove console

File: src/utils/methods.js [hunk 20]
```diff
@@ -1,90 +1,125 @@
...
+    if (utils.getCurrentTime() - new Date(lastUpdated) > REFRESH_TIME) { //refresh time limit
+      // console.log("gettin position");
```
Comment (rich521) [inline src/utils/methods.js]: remove

File: src/utils/methods.js [hunk 21]
```diff
@@ -1,90 +1,125 @@
...
       // check each item, then refetch if needed
       position = await utils.getCurrentPosition();
-      weather = await utils.getCurrentWeather(position);
+      // console.log("gettin weather");
```
Comment (rich521) [inline src/utils/methods.js]: remove

File: src/utils/methods.js [hunk 22]
```diff
@@ -1,90 +1,125 @@
...
+      weather = await utils.getCurrentWeather(position.coords);
+      console.log("processing weather data");
+      let { description, isRaining } = await utils.retrieveDayForecast(weather,weather.cnt);
+      console.log("resetting last updated weather");
```
Comment (rich521) [inline src/utils/methods.js]: remove

File: src/utils/methods.js [hunk 23]
```diff
@@ -1,90 +1,125 @@
...
         resolve(response.data);
       });
     });
-     //return(null);
+    //return(null);
+  },
+
+  retrieveDayForecast: async (data,count) => {
+    const d = new Date()
+    const z = n => n.toString().length == 1 ? `0${n}` : n // Zero pad
```
Comment (rich521) [inline src/utils/methods.js]: what is this?

Comment (rich521) [reply src/utils/methods.js]: use this for hard === (includes type)

File: src/utils/methods.js [hunk 24]
```diff
@@ -1,90 +1,125 @@
...
   },
 
-  getCachedSettings: async () => {
-    // Get the local settings
-    const localStore = await utils.getLocalData(KEY.SETTINGS);
-
+  getCachedItems: async () => {
+    // Get the localdata
+    const localStore = await utils.getLocalData(KEY.WEATHER);
+    console.log(localStore);
```
Comment (rich521) [inline src/utils/methods.js]: remove

File: src/utils/methods.js [hunk 25]
```diff
@@ -1,90 +1,125 @@
...
+    const z = n => n.toString().length == 1 ? `0${n}` : n // Zero pad
+    const date = `${d.getFullYear()}-${z(d.getMonth()+1)}-${z(d.getDate())}` // returns (2017-11-10 ie YYYY-MM-DD)
+
+    let description = "";
+    for (let i=0; i< count; i++){
+      if (data.list[i].dt_txt.indexOf(date) === -1) break; //find if the items in the list are todays forcast only
+      description = `${description} ${data.list[i].weather[0].description}`; //append descriptions into one variable
+    }
+    let isRaining = false;
+    if( description.indexOf("rain") >0 ) isRaining = true;
```
Comment (rich521) [inline src/utils/methods.js]: 0 --> also means it exists. do you mean ```> -1 ```

File: package.json [hunk 26]
```diff
@@ -7,9 +7,11 @@
 		"test": "jest"
 	},
 	"dependencies": {
-		"axios": "^0.16.2",
+		"axios": "0.16.2",
 		"react": "16.0.0-beta.5",
 		"react-native": "0.49.3",
+		"react-native-background-task": "0.2.1",
+		"react-native-push-notification": "3.0.1",
 		"react-native-router-flux": "^4.0.0-beta.22"
```
Comment (rich521) [inline package.json]: remove ^

File: src/App.js [hunk 27]
```diff
@@ -1,25 +1,115 @@
...
 // import { PermissionsAndroid } from 'react-native';
 import utils from './utils/methods';
-import defaultState from './utils/constants';
+import { KEY } from './utils/constants';
+
+const TASK_PERIOD = 900;
+
+BackgroundTask.define(async () => {
+  const ID = 'ID: ' + Math.random().toFixed(4);
+  console.log(`1- ${ID} - ${new Date()} Background Job fired!.`);
```
Comment (rich521) [inline src/App.js]: remove lines 15, 16

File: src/App.js [hunk 28]
```diff
@@ -1,25 +1,115 @@
...
+    utils.getCachedItems().then(data => {
+      this.setState({ ...data , remark:true});
+    });
+
+    if (PushNotification) PushNotification.cancelAllLocalNotifications();
+    PushNotification.localNotification({
+      title: "No umbrella needed.",
+      largeIcon: "icon",
+      smallIcon: "icon",
+      message: "Enjoy your day ;)", // (required)
```
Comment (rich521) [inline src/App.js:72]: add TODO

File: src/App.js [hunk 29]
```diff
@@ -1,25 +1,115 @@
...
+     AppState.addEventListener('change', this.handleAppStateChange);
+  }
+
+  handleAppStateChange = async (appState) => {
+    if (appState === 'background') {
+      this.setState({ isRaining: false });
+      const settings = await utils.fetchSettings();
+      const period_difference = new Date(new Date(settings.date) - Date.now());
+      BackgroundTask.schedule({
+          period: (period_difference.getHours()*60*60 + period_difference.getMinutes()*60), //calculate time to set (s)
```
Comment (rich521) [inline src/App.js]: ```(period_difference.getHours()*60*60 + period_difference.getMinutes()*60)``` store as variable and describe what it does in the variable name

File: src/App.js [hunk 30]
```diff
@@ -1,25 +1,115 @@
...
+      });
+    }
+    if (appState === 'active') {
+      BackgroundTask.cancel();
+      utils.getCachedItems().then(data => {
+        this.setState({ ...data });
+        console.log("recieved cached items");
```
Comment (rich521) [inline src/App.js]: remove this

File: src/App.js (hunk 29 above)
Comment (rich521) [reply src/App.js]: ```new Date(settings.date) => remindDate```

Comment (rich521) [reply src/App.js]: remindLaterTimeInSeconds

File: src/Settings.js [hunk 31]
```diff
@@ -1,43 +1,87 @@
-import React from 'react';
-import { View, Text } from 'react-native';
-import { Card, CardSection, Button } from './components/common';
+import React, { Component } from 'react';
+import { View, Text, Switch } from 'react-native';
 import { Actions } from 'react-native-router-flux';
+import { Card, CardSection, Button } from './components/common';
+import DateModal from './components/DateModal';
+//import { SCENE } from './utils/constants';
```
Comment (rich521) [inline src/Settings.js]: remove this

**Review by rich521 — APPROVED** [review]

**Comment by rich521** [comment]
+1

**Pull request merged**